apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -0,0 +1,408 @@
+    perform_connected_healthcheck(Data);
+handle_event(enter, _OldState, disconnected, #data{id = InstId} = Data) ->
+    UpdatedData = Data#data{status = disconnected},
+    ets:delete(?ETS_TABLE, InstId),
+    {next_state, disconnected, UpdatedData}.
+
+%%------------------------------------------------------------------------------
+%% internal functions
+%%------------------------------------------------------------------------------
+
+do_start(InstId, Group, ResourceType, Config, Opts) ->
+    % The state machine will make the actual call to the callback/resource module after init
+    emqx_resource_manager_sup:start_child(InstId, Group, ResourceType, Config, Opts),
+    case wait_for_resource_ready(InstId, maps:get(wait_for_resource_ready, Opts, 5000)) of
+        ok ->
+            ok = emqx_plugin_libs_metrics:create_metrics(
+                resource_metrics,
+                InstId,
+                [matched, success, failed, exception],
+                [matched]
+            ),
+            ok;
+        timeout ->
+            {error, timeout}
+    end.
+
+do_stop(Data) ->
+    Result = emqx_resource:call_stop(Data#data.id, Data#data.mod, Data#data.state),
+    ets:delete(?ETS_TABLE, Data#data.id),
+    Result.
+
+proc_name(Id) ->
+    binary_to_atom(Id).
+
+connecting_healthcheck(Data) ->
+    case emqx_resource:call_health_check(Data#data.id, Data#data.mod, Data#data.state) of
+        {ok, ResourceState} ->
+            UpdatedData = Data#data{state = ResourceState, status = connected, error = undefined},
+            ets:insert(?ETS_TABLE, {Data#data.id, Data#data.group, UpdatedData}),
+            {next_state, connected, UpdatedData};
+        {error, Reason} ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, Reason]),
+            UpdatedData = Data#data{error = Reason},
+            Actions = [{state_timeout, ?SHORT_HEALTHCHECK_INTERVAL, healthcheck}],
+            {keep_state, UpdatedData, Actions};
+        {error, Reason, ResourceState} ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, Reason]),
+            UpdatedData = Data#data{state = ResourceState, error = Reason},
+            Actions = [{state_timeout, ?SHORT_HEALTHCHECK_INTERVAL, healthcheck}],
+            {keep_state, UpdatedData, Actions}
+    end.
+
+perform_connected_healthcheck(Data) ->
+    case emqx_resource:call_health_check(Data#data.id, Data#data.mod, Data#data.state) of
+        {ok, ResourceState} ->
+            UpdatedData = Data#data{state = ResourceState},
+            ets:insert(?ETS_TABLE, {Data#data.id, Data#data.group, UpdatedData}),
+            Actions = [{state_timeout, ?HEALTHCHECK_INTERVAL, healthcheck}],
+            {keep_state, UpdatedData, Actions};
+        {error, Reason} ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, Reason]),
+            UpdatedData = Data#data{error = Reason},
+            ets:delete(?ETS_TABLE, Data#data.id),
```

Comment:
For my own information: would there be a problem to keep the ETS row for instances in other states, since the state is contained in the third element?

Reply:
Thought I had answered this on Friday. No, there isn't actually an issue with it. That is exactly how things operated previously. The design doc specifically called out removing the ETS state if the state machine state had moved into anything but a `connected` state.

What that does do though, as you commented on earlier, is remove the ability to respond with a clear reason why you can't query as we don't have the state information anymore to perform the case check on to give a more detailed reason. That is something that could be put back in.